web/packages/core/src/i18n.ts

```diff
@@ -0,0 +1,123 @@
+
+for (const locale in BUNDLED_TEXTS) {
+    const bundle = new FluentBundle(locale);
+    const files = BUNDLED_TEXTS[locale];
+    if (files) {
+        for (const filename in files) {
+            const text = files[filename];
+            if (text) {
+                for (const error in bundle.addResource(
+                    new FluentResource(text)
+                )) {
+                    console.error(
+                        `Error in text for ${locale} ${filename}: ${error}`
+                    );
+                }
+            }
+        }
+    }
+    bundles[locale] = bundle;
+}
+
+/**
+ * Gets the localised text for the given locale and text ID.
+ *
+ * If the locale does not contain a text for this ID, it will return null.
+ *
+ * @param locale Locale to prefer when retrieving text, ie "en-US"
+ * @param id ID of the text to retrieve
+ * @param args Any arguments to use when creating the localised text
+ * @returns Localised text or null if not found
+ */
+function tryText(
+    locale: string,
+    id: string,
+    args?: Record<string, FluentVariable> | null
+): string | null {
+    const bundle = bundles[locale];
+    if (bundle !== undefined) {
+        const message = bundle.getMessage(id);
+        if (message !== undefined && message.value) {
+            return bundle.formatPattern(message.value, args);
+        }
+    }
+    return null;
+}
+
+/**
+ * Gets the localised text for the given text ID.
+ *
+ * The users preferred locales are used, in priority order, to find the given text.
+ *
+ * If no text is found for any preferred locale, en-US will be used.
+ * If en-US does not contain a text for this ID, an error will be logged and the ID itself will be returned.
+ *
+ * @param id ID of the text to retrieve
+ * @param args Any arguments to use when creating the localised text
+ * @returns Localised text
+ */
+export function text(
+    id: string,
+    args?: Record<string, FluentVariable> | null
+): string {
+    const locales = negotiateLanguages(
+        navigator.languages,
+        Object.keys(bundles),
+        { defaultLocale: "en-US" }
+    );
+
+    for (const i in locales) {
+        const result = tryText(locales[i]!, id, args);
+        if (result) {
+            return result;
+        }
+    }
+
+    console.error(`Unknown text key '${id}'`);
+    return id;
+}
+
+/**
+ * Gets the localised text for the given text ID, as <p>paragraphs</p> and HTML entities safely encoded.
+ *
+ * The users preferred locales are used, in priority order, to find the given text.
+ *
+ * If no text is found for any preferred locale, en-US will be used.
+ * If en-US does not contain a text for this ID, an error will be logged and the ID itself will be returned.
+ *
+ * @param id ID of the text to retrieve
+ * @param args Any arguments to use when creating the localised text
+ * @returns Localised text with each line in a Paragraph element
+ */
+export function textAsParagraphs(
```

Comment:
I have not looked at the code closely, and this seems like an improvement on the existing code, but I do not really like the general usage of assigning to `innerHTML`, like here: https://github.com/ruffle-rs/ruffle/blob/master/web/packages/core/src/ruffle-player.ts#L1890 . Important to note is that it is the existing code, not your PR, that uses this `innerHTML` . So your PR is not really introducing anything. But I still do not like it. I do not know how translations are gotten into the project, if they are committed to the repository or how - if translations are committed to the repository, they will have gone through review either way. Your way of doing it does seem fine, since it is not introducing anything that is not already being done, and your code does seem like an improvement. But I would not be sad if people make an issue or other PRs to change away from using `innerHTML` in 'ruffle-player.ts' and elsewhere.

Would it make sense for me to make an issue that describes that refactoring away from using `innerHTML` in 'web/packages/core/src/ruffle-player.ts' for translations (and other usage) would be nice (not required)? Or is there already a TODO for it in the code?

I am tempted to propose that you add a new function, `textAsParagraphs`, with return type `HTMLElement`, and rename this function `textAsParagraphsInnerHTML`, and then call `textAsParagraphsInnerHTML` from `textAsParagraphs`. Simply such that other users have the choice of using the `HTMLElement`-returning or using the `string`-returning variant.

I do not consider this a blocker, and I will let it be up to you what to do or not change anything in the PR, this is mostly me feeling that `innerHTML` is somewhat of a code smell.

Reply:
Would it make sense for me to make an issue on Github for cleaning up the `innerHTML`-usage in 'web/packages/core/src/ruffle-player.ts' and elsewhere?

Reply:
It doesn't make sense to me to have a method that returns a div when everything currently wants a html string of <p>s

Reply:
That makes sense, it was more for future usage, in case it is refactored later to not use `innerHTML` - but it is easy to add such a function later. I saw it more as changing the internal "API" a little bit.

Reply:
I have created https://github.com/ruffle-rs/ruffle/issues/10976 .

(I think all of my questions have been answered, please do mark the review lines here resolved as you see fit).